### ISSUE TITLE
How to get all claims for a reference token?

### ISSUE THREAD

**Issue by urbanhusky** [IdentityServer/IdentityServer3#2313]
Hi,

I'm using reference tokens and a mix of resource owner credentials flow, client credentials flow and a custom grant. The application is an ASP.NET MVC/WebApi with WCF services.

Communication, authentication and authorization work fine so far - except that I cannot retrieve all claims from the token in my WCF services. I'm wrapping the token and handle it in a `SecurityTokenHandler`, where I try to set the claims identity.

My issue is as follows:
I cannot query the `UserInfo` endpoint for clients that are not users (i.e. client credentials flow) but I do not know that until I try. This way I log many error messages because the tokens from the client credentials flow cannot have the openid scope and therefore cause to log `ERROR; IdentityServer3.Core.Validation.TokenValidator; Checking for expected scope openid failed`

Querying the token introspection endpoint does not yield the user claims - and I face an additional problem: I only the the claims for the scope that I authenticate with, but I can't know which scopes the client requested. 
I currently solve this by defining a superset-scope over all scopes, which is always present. 
I still won't know the exact scopes but at least I get all the claims I need. This kind of defeats partitioning the claims in the first place, except in the token validation middlewares that only query their respective scopes...

The only alternative I could think of would be to query the introspection endpoint for each and every possible scope in my application (on each request...).

tl;dr: Userinfo not applicable for client credentials flow but I can't know that I shouldn't query it from just a reference token alone. Can't get user claims (such as first name etc.) from introspection endpoint?

**Comment by leastprivilege** [comment]
"I'm using reference tokens and a mix of resource owner credentials flow, client credentials flow and a custom grant."

Maybe that is the root cause of your problem.

wrt to introspection - I added the ability to introspect all scope claims. the version is on the dev branch and the myget feed.

**Comment by pashute** [comment]
You need a reference type scope with the claimtypes defined in it, and you need to ask for that scope in the middleware as well as in the original call

**Comment by urbanhusky** [comment]
Defining a resource scope that has these missing claims does not yield the user-claims. Being able to define a scope that can introspect all other scopes might help, I'll test that when it's released :)

Right now I've got a workaround of defining a "HasUserClaims" claim on the clients, which I then use to determine if I can, theoretically, safely query the userinfo endpoint.

**Comment by pashute** [comment]
It did not yield the user claims, but it did show those claims inside the access_token given to my API as a bearer token, when asking for that scope in the WinAPI project's startup. I then read those manually inside my custom Authorize attribute. 

I hope I'm not mixing in and just complicating things for urbanhusky or Dominick who is the expert. 

I'll stay out of the discussion if you ask me to.
